runtime/include/cloe/data_broker.hpp

```diff
@@ -0,0 +1,1099 @@
+  return *lhs < *rhs;
+}
+template <typename T, typename U>
+constexpr bool operator<=(const BasicContainer<T>& lhs, const BasicContainer<U>& rhs) {
+  return *lhs <= *rhs;
+}
+template <typename T, typename U>
+constexpr bool operator>(const BasicContainer<T>& lhs, const BasicContainer<U>& rhs) {
+  return *lhs > *rhs;
+}
+template <typename T, typename U>
+constexpr bool operator>=(const BasicContainer<T>& lhs, const BasicContainer<U>& rhs) {
+  return *lhs >= *rhs;
+}
+
+// Compare BasicContainer with a value
+
+template <typename T, typename U>
+constexpr bool operator==(const BasicContainer<T>& lhs, const U& rhs) {
+  return *lhs == rhs;
+}
+template <typename T, typename U>
+constexpr bool operator==(const T& lhs, const BasicContainer<U>& rhs) {
+  return lhs == *rhs;
+}
+template <typename T, typename U>
+constexpr bool operator!=(const BasicContainer<T>& lhs, const U& rhs) {
+  return *lhs != rhs;
+}
+template <typename T, typename U>
+constexpr bool operator!=(const T& lhs, const BasicContainer<U>& rhs) {
+  return lhs != *rhs;
+}
+template <typename T, typename U>
+constexpr bool operator<(const BasicContainer<T>& lhs, const U& rhs) {
+  return *lhs < rhs;
+}
+template <typename T, typename U>
+constexpr bool operator<(const T& lhs, const BasicContainer<U>& rhs) {
+  return lhs < *rhs;
+}
+template <typename T, typename U>
+constexpr bool operator<=(const BasicContainer<T>& lhs, const U& rhs) {
+  return *lhs <= rhs;
+}
+template <typename T, typename U>
+constexpr bool operator<=(const T& lhs, const BasicContainer<U>& rhs) {
+  return lhs <= *rhs;
+}
+template <typename T, typename U>
+constexpr bool operator>(const BasicContainer<T>& lhs, const U& rhs) {
+  return *lhs > rhs;
+}
+template <typename T, typename U>
+constexpr bool operator>(const T& lhs, const BasicContainer<U>& rhs) {
+  return lhs > *rhs;
+}
+template <typename T, typename U>
+constexpr bool operator>=(const BasicContainer<T>& lhs, const U& rhs) {
+  return *lhs >= rhs;
+}
+template <typename T, typename U>
+constexpr bool operator>=(const T& lhs, const BasicContainer<U>& rhs) {
+  return lhs >= *rhs;
+}
+
+/**
+ * Signal-Descriptor
+ * \note: Design-Goals:
+ * - Design-#1: The class shall expose a uniform interface (via type-erasure)
+ * - Design-#2: The class shall be constructible via external helpers
+ * - Design-#3: CppCoreGuidelines CP-1 does not apply. See also: https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rconc-multi
+ * - Design-#4: Templates shall be instantiated explicitly
+ * \note: Implementation-Notes:
+ * - Implementation-#1: Type specific aspects are implemented in templates (Design-#1)
+ * - Implementation-#2: Objects are created via a factory-method (to prevent instances of incomplete initialization)
+ * - Implementation-#3: An access-token is used to have a public c'tor which in fact shall be inaccesible (Design-#2).
+ */
+class Signal {
+ public:
+  // Signal types
+  using signal_ptr_t = std::shared_ptr<Signal>;
```

Comment:
Since this is a type that will be used by users, I'd like to have:
- it exist outside `Signal`, as `using SignalPtr = std::shared_ptr<Signal>`
- Signal doesn't need to know what our preferred smart pointer type is

The factory method `create()` below should return a `std::unqiue_ptr<Signal>`, since this makes clear that the caller is the new owner. This is cheap and can be transparently converted to a `std::shared_ptr<Signal>` by the data broker.

Reply:
I admit I find it difficult, especially since there's a certain clash when it comes to templates. We have for example, the `is_compatible_t`, which I find aligns well with all the other template functions we use, and is orthogonal with the STL, and that value is great enough that I feel it is worth going against the convention of camelcase for types.

Same thing for type using statements.